File: ANZAC 4 2023/k_incomplete/K.cpp
// Fails with RUN-ERROR

#include <iostream>
#include <vector>
#include <unordered_set>

using namespace std;

int main() {
    size_t n, q;
    cin >> n >> q;

    vector<size_t> str(n);
    vector<unordered_set<size_t>> indices(n);
    for (size_t i = 0; i < n; i++) {
        str[i] = i;
        indices[i].insert(i);
    }

    while (q--) {
        int type;
        cin >> type;
        if (type == 1) {
            size_t l, r;
            cin >> l >> r;
            l--; r--;
            for (size_t i = 0; i <= (r - l) / 2; i++) {
                const size_t& old_char = str[r - i];
                const size_t& new_char = str[l + i];
                if (old_char == new_char) continue;

                auto& old_idx = indices[old_char];
                auto& new_idx = indices[new_char];
                new_idx.insert(old_idx.begin(), old_idx.end());
                for (size_t idx : old_idx) {
                    str[idx] = new_char;
                }
            }
        } else {
            size_t a, b, x, y;
            cin >> a >> b >> x >> y;
            a--; b--; x--; y--;
            if (b - a != y - x) {
                cout << "Not equal\n";
                continue;
            }
            bool eq = true;
            for (size_t i = 0; i <= (b - a); i++) {
                if (str[a + i] != str[x + i]) {
                    eq = false;
                    break;
                }
            }
            if (eq) cout << "Equal\n";
            else cout << "Unknown\n";
        }
    }
}

File: ANZAC 4 2023/k_incomplete/K.py
# Fails with WRONG-ANSWER

n, q = (int(n) for n in input().split())
string = list(range(n))
indices = [{i} for i in range(n)]

for _ in range(q):
    query, *args = (int(n) - 1 for n in input().split())
    if query == 0:
        l, r = args
        for i in range(0, (r - l + 1) // 2):
            l_char = string[l + i]
            r_char = string[r - i]
            if l_char == r_char: continue

            r_indices = indices[r - i]
            indices[l + i].update(r_indices)
            for idx in r_indices:
                string[idx] = l_char
        print(string)
    else:
        a, b, x, y = args
        range1 = range(a, b + 1)
        range2 = range(x, y + 1)
        if len(range1) != len(range2):
            print('Not equal')
            continue

        equal = True
        for a, b in zip(range1, range2):
            if string[a] != string[b]:
                equal = False
                break

        if equal:
            print('Equal')
        else:
            print('Unknown')
